### PR TITLE
Fix race condition on static dictionaries

### PR DESCRIPTION
**Pull request by udlose**
*Issue #282 *

*Description of changes:*
- fix race condition on static dictionaries, copy recent code from LitJson repo for JsonMapper.cs

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jj22ee** [comment]
```
Failed TestExceptionStrategy7 [6 ms]
  Error Message:
   Test method Amazon.XRay.Recorder.UnitTests.AwsXrayRecorderBuilderTests.TestExceptionStrategy7 threw exception: 
ThirdParty.LitJson.JsonException: Max allowed object depth reached while trying to export from type System.String
  Stack Trace:
      at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 715
```

Looks like all failing tests are due to a recursive (infinite?) loop at the [WriteValue()](https://github.com/aws/aws-xray-sdk-dotnet/blob/3a5cf9906609322648dedf8c92e5e0f81ac6e3f3/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs#L710-L712) method.
```
ThirdParty.LitJson.JsonException: Max allowed object depth reached while trying to export from type System.String
  Stack Trace:
      at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 715
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 792
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 850
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 850
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 783
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 850
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 850
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 783
   at ThirdParty.LitJson.JsonMapper.WriteValue(Object obj, JsonWriter writer, Boolean writer_is_private, Int32 depth) in /_/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs:line 850
...
```

Can you fix the PR, and then we will review the change.

**Comment by udlose** [comment]
@jj22ee my apologies.  I had a PITA time trying to get it to build successfully with all of the old dependencies used for backwards compatibility. So, I wasn't able to run any of the tests. I will try to take a look this weekend.

Do you have any tips on making it a little easier to build?

**Comment by jj22ee** [comment]
@udlose are there any specific dependencies you are having trouble with, and what kind of build pains are you running into?

**Comment by udlose** [comment]
> @udlose are there any specific dependencies you are having trouble with, and what kind of build pains are you running into?

The .NET Framework specific dependencies are problematic as I keep only the latest version installed on my laptop.  I went and installed .NET Fx 4.62 and .Net Fx 4.51. There are multiple references to the compilation symbol NET45 which apparently aren't covered by installing .NET Framework 4.51.  I pulled NET45 out of the project dependencies to try and get it to build but there are several references to it in the code for conditional compilation.  I don't want to have to do surgery on the code just to get it to compile correctly.

**Comment by jj22ee** [comment]
I have pushed a fix into the PR branch:
1. There was some deleted code in your PR that I restored which does not exist in upstream LitJSON.
2. Regarding the other synced changes, I removed them as these were incomplete since they did not incorporate the other changes in their respective commits:
    - https://github.com/LitJSON/litjson/commit/1a120dff7e9b677633bc568322fa065c9dfe4bb8
    - https://github.com/LitJSON/litjson/commit/726c05debc9bb8c4a3a57ff0aabd843d1b26aba0
    - https://github.com/LitJSON/litjson/commit/4828d2eee5b23b3596b52a4b57bf4b5d37c3b5c3
    - https://github.com/LitJSON/litjson/commit/e5bbcc8ea6040c3b8cac0539b05b2c234bac2517

To summarize, I've reduced the PR into essentially cherry-picking [your upstream PR](https://github.com/LitJSON/litjson/pull/142/files), alongside incorporating your change into the other existing code change to the copied LitJSON.

**Review by jj22ee — APPROVED** [review]
Tests now pass, LGTM as it now essentially a cherry-pick of https://github.com/LitJSON/litjson/pull/142. Let me know if you have concerns with my update to your PR.
